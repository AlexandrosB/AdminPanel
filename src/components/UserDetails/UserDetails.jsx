import React, { Component } from 'react';
import PropTypes from 'prop-types';

require('./Static/userDetails.scss');

class UserDetails extends Component {
    static propTypes = {
        id: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        email: PropTypes.string.isRequired,
        phone: PropTypes.string.isRequired,
        address: PropTypes.string.isRequired,
        company: PropTypes.string.isRequired,
        updateUserDetails: PropTypes.func.isRequired
    };

    constructor(props) {
        super(props);
        const {
            name, email, phone, address, company
        } = this.props;
        this.state = {
            name,
            email,
            phone,
            address,
            company
        };
    }

    componentDidUpdate(oldProps) {
        const newProps = this.props;
        if (oldProps.id !== newProps.id) {
            this.setState({
                name: newProps.name,
                email: newProps.email,
                phone: newProps.phone,
                address: newProps.address,
                company: newProps.company
            });
        }
    }

    /**
     * Change the value of
     * the selected field
     * @param event
     * @param type
     */
    updateFields = (event, type) => {
        this.setState({
            [type]: event.target.value
        });
    };

    /**
     * Cancel all the changes
     * and restore the original
     * values
     */
    cancelChanges = () => {
        const {
            name, email, phone, address, company
        } = this.props;
        this.setState({
            name,
            email,
            phone,
            address,
            company
        });
    };

    /**
     * Update user info with
     * the new values
     */
    updateUserDetails = () => {
        const {
                name, email, phone, address, company
            } = this.state,
            { updateUserDetails, id, photo } = this.props,
            data = {
                id,
                photo,
                name,
                company,
                email,
                phone,
                address
            };
        updateUserDetails(data);
    };

    render() {
        const {
                name, email, phone, address, company
            } = this.state,
            {
                name: oldName, email: oldEmail, phone: oldPhone, address: oldAddress, company: oldCompany
            } = this.props,
            enableButton = (name !== oldName || email !== oldEmail
                || phone !== oldPhone || address !== oldAddress
                || company !== oldCompany);
        return (
            <section className='userDetailsContainer'>
                <div className='formGroup'>
                    <div className='formElement'>
                        <span className='label'>Name</span>
                        <input
                            type='text'
                            value={ name }
                            placeholder='Enter name'
                            onChange={ (e) => this.updateFields(e, 'name') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Email address</span>
                        <input
                            type='text'
                            value={ email }
                            placeholder='Enter email address'
                            onChange={ (e) => this.updateFields(e, 'email') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Phone</span>
                        <input
                            type='text'
                            value={ phone }
                            placeholder='Enter phone'
                            onChange={ (e) => this.updateFields(e, 'phone') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Address</span>
                        <input
                            type='text'
                            value={ address }
                            placeholder='Enter address'
                            onChange={ (e) => this.updateFields(e, 'address') }
                        />
                    </div>
                    <div className='formElement'>
                        <span className='label'>Company</span>
                        <input
                            type='text'
                            value={ company }
                            placeholder='Enter company'
                            onChange={ (e) => this.updateFields(e, 'company') }
                        />
                    </div>
                    <section className='buttons'>
                        <button
                            className='cancel'
                            onClick={ this.cancelChanges }
                            disabled={ !enableButton }
                        >
                            <p>Cancel</p>
                        </button>
                        <button
                            className='update'
                            onClick={ this.updateUserDetails }
                            disabled={ !enableButton }
                        >
                            <p>Save</p>
                        </button>
                    </section>
                </div>
            </section>
        );
    }
}

export default UserDetails;
