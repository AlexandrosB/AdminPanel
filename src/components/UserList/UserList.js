import React, { Component } from 'react';
import PropTypes from 'prop-types';
import UserListSingleUser from './Static/UserListSingleUser';

require('./Static/userList.scss');

class UserList extends Component {
    render() {
        const { users, selectUser, selectedUserId } = this.props;
        return (
            <ul className='userListContainer'>
                {
                    users.map((user) => (
                        <UserListSingleUser
                            key={ user.id }
                            userId={ user.id }
                            userPhoto={ user.photo }
                            userName={ user.name }
                            userEmail={ user.email }
                            selectUser={ selectUser }
                            selectedUserId={ selectedUserId }
                        />
                    ))
                }
            </ul>
        );
    }
}

UserList.propTypes = {
    users: PropTypes.array.isRequired,
    selectUser: PropTypes.func.isRequired,
    selectedUserId: PropTypes.string.isRequired
};

export default UserList;
