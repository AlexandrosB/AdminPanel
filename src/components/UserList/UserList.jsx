import React, { Component } from 'react';
import PropTypes from 'prop-types';
import UserListSingleUser from './Static/UserListSingleUser';

require('./Static/userList.scss');

class UserList extends Component {
    static propTypes = {
        users: PropTypes.array.isRequired,
        selectUser: PropTypes.func.isRequired,
        selectedUserId: PropTypes.string.isRequired
    };

    render() {
        const { users, selectUser, selectedUserId } = this.props;
        return (
            <ul className='userListContainer'>
                {
                    users.map((user) => (
                        <UserListSingleUser
                            key={ user.id }
                            userId={ user.id }
                            userPhoto={ user.photo }
                            userName={ user.name }
                            userEmail={ user.email }
                            selectUser={ selectUser }
                            selectedUserId={ selectedUserId }
                        />
                    ))
                }
            </ul>
        );
    }
}

export default UserList;
