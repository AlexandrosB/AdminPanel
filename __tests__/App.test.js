import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { USERS } from '../__testData__/testData';

import App from '../src/App';
import UserList from '../src/components/UserList/UserList';
import UserDetails from '../src/components/UserDetails/UserDetails';

describe('<App />', () => {
    const selectUser = jest.fn(),
        updateUserDetails = jest.fn(),
        wrapper = shallow(<App
            selectUser={ selectUser }
            updateUserDetails={ updateUserDetails }
        />);

    it('should match the snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('renders without crashing', () => {
        wrapper;
    });

    it('renders a section that contains child components', () => {
        expect(wrapper.find('section.appContainer').exists()).toBe(true);
    });

    it('contains UserList component', () => {
        expect(wrapper.find(UserList).exists()).toBe(true);
    });

    it('contains UserDetails component', () => {
        const secondaryWrapper = shallow(<App selectedUser={ USERS[0] } />);
        const element = secondaryWrapper.find(UserDetails);
        expect(element).toBeTruthy();
    });

    it('does not contain UserDetails component', () => {
        const secondaryWrapper = shallow(<App selectedUser={ null } />);
        const element = secondaryWrapper.find('div.noUser');
        expect(element).toBeTruthy();
    });

    it('check state values', () => {
        expect(wrapper.state('selectedUser')).toBe(null);
        expect(wrapper.state('users')).toBeArray();
    });
});
