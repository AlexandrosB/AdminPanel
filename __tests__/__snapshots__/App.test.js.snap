// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should match the snapshot 1`] = `
<section
  className="appContainer"
>
  <UserList
    selectUser={[Function]}
    selectedUserId=""
    users={
      Array [
        Object {
          "address": "958 Brevoort Place, Ona, Maine, 2433",
          "company": "ZOLAREX",
          "email": "bates.washington@zolarex.io",
          "id": "5c093af1c6ee9117a581c7d6",
          "name": "Bates Washington",
          "phone": "+1 (915) 447-2207",
          "photo": "https://randomuser.me/api/portraits/men/40.jpg",
        },
        Object {
          "address": "120 Cedar Street, Mansfield, Kentucky, 8890",
          "company": "VIAGRAND",
          "email": "mollie.oneill@viagrand.biz",
          "id": "5c093af1aeca1bb00607fb2a",
          "name": "Mollie Oneill",
          "phone": "+1 (852) 535-3880",
          "photo": "https://randomuser.me/api/portraits/men/1.jpg",
        },
        Object {
          "address": "673 Revere Place, Canoochee, New Mexico, 4980",
          "company": "SPEEDBOLT",
          "email": "",
          "id": "5c093af11dc4289c2941bd83",
          "name": "Lana Sharpe",
          "phone": "+1 (876) 475-3370",
          "photo": "https://randomuser.me/api/portraits/men/2.jpg",
        },
        Object {
          "address": "716 Hazel Court, Hanover, Louisiana, 9603",
          "company": "STOCKPOST",
          "email": "massey.ewing@stockpost.ca",
          "id": "5c093af172c42a579bdb685f",
          "name": "Massey Ewing",
          "phone": "+1 (948) 482-2392",
          "photo": "https://randomuser.me/api/portraits/men/3.jpg",
        },
        Object {
          "address": "371 Cortelyou Road, Farmers, Iowa, 1733",
          "company": "DIGIAL",
          "email": "patrica.melton@digial.name",
          "id": "5c093af1264dbf1a467091cf",
          "name": "Patrica Melton",
          "phone": "",
          "photo": "https://randomuser.me/api/portraits/men/4.jpg",
        },
        Object {
          "address": "787 Bleecker Street, Nipinnawasee, New York, 6872",
          "company": "ENTALITY",
          "email": "",
          "id": "5c093af1e39cab35f1c9726e",
          "name": "Viola Serrano",
          "phone": "+1 (909) 489-2270",
          "photo": "https://randomuser.me/api/portraits/men/5.jpg",
        },
        Object {
          "address": "293 Linwood Street, Roulette, Nevada, 7805",
          "company": "XTH",
          "email": "tammy.dixon@xth.net",
          "id": "5c093af1dd7402c939bd490c",
          "name": "Tammy Dixon",
          "phone": "+1 (810) 472-2762",
          "photo": "https://randomuser.me/api/portraits/men/6.jpg",
        },
        Object {
          "address": "600 Calder Place, Bourg, Hawaii, 9397",
          "company": "SKYPLEX",
          "email": "christensen.brown@skyplex.us",
          "id": "5c093af15b557da4238bcb0c",
          "name": "Christensen Brown",
          "phone": "+1 (892) 462-2141",
          "photo": "https://randomuser.me/api/portraits/men/7.jpg",
        },
        Object {
          "address": "401 Spaight Way",
          "company": "Dabvine",
          "email": "wberends0@unblog.fr",
          "id": "5c091af1e39cab35f9c9726n",
          "name": "Wilmer Berends",
          "phone": "111-757-1393",
          "photo": "https://randomuser.me/api/portraits/men/8.jpg",
        },
        Object {
          "address": "1 Colorado Junction",
          "company": "Fivespan",
          "email": "yculver1@netvibes.com",
          "id": "5c693af1dd7402c939bd4806",
          "name": "Yolane Culver",
          "phone": "471-485-5905",
          "photo": "https://randomuser.me/api/portraits/men/9.jpg",
        },
        Object {
          "address": "6977 Sugar Trail",
          "company": "Shuffledrive",
          "email": "cmarthen2@alibaba.com",
          "id": "50093af15b557da4248bcb02",
          "name": "Clarance Marthen",
          "phone": "638-994-3228",
          "photo": "https://randomuser.me/api/portraits/men/10.jpg",
        },
      ]
    }
  />
  <div
    className="noUser"
  >
    Select a user!
  </div>
</section>
`;
